### PR TITLE
[Helpers] Move towards feature parity and reliability improvements

### LINKED ISSUE #896: Mochawesome helper does not work with shell command
#### What are you trying to achieve?
I'm trying to run the shell command

#### What do you get instead?
The shell command is not working.

> Provide console output if related. Use `--verbose` mode for more details.

```bash
 ./node_modules/codeceptjs/bin/codecept.js shell --verbose
String interactive shell for current suite...
   [1] Starting recording promises
   Emitted | suite.before ([object Object])
   [1] Queued | hook Nightmare._beforeSuite()
   [1] Queued | hook Mochawesome._beforeSuite()
   Emitted | test.before
   [1] Queued | hook Nightmare._before()
   [1] Queued | hook Mochawesome._before()
   [1] Queued | Start new session
   [1] Queued | () => event.emit(event.test.after)
   [1] Queued | () => event.emit(event.suite.after, {})
   [1] Queued | () => codecept.teardown()
 > [Session] Starting singleton browser session
(node:15157) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): TypeError: Cannot read property 'currentTest' of undefined
(node:15157) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

### Details

* CodeceptJS version: 1.1.2
* NodeJS Version: 8.9.4
* Operating System: OSX
* Protractor || WebDriverIO || Nightmare version (if related)
* Configuration file:
```"helpers": {
    "Nightmare": {
    },
    "Mochawesome": {
      "uniqueScreenshotNames": "true"
    }
  }
```

The problem appears to be _beforeSuite(suite) is getting suite = {} (empty object)

### PR DESCRIPTION
**Pull request by reubenmiller**
Apologies in advance for the size of this pull request, I have been focusing on feature parity between the helpers, but had to make some other changes which inpacted the running of my tests. Since some of the tests for the helpers are shared, I wasn't able to quickly seperate the different commits easily.

However let me know if I need to split the PR (and if so how).

### Steps towards feature parity in Web Helpers (i.e. Puppeteer/Protractor/WebDriverIO/Nightmare)

This PR focuses on moving towards feature parity between the Web Helpers. I have created a table with the overview of the support functions across the helpers. The new functions are highlighted in purple.

[Helper API Overview](https://docs.google.com/spreadsheets/d/1MrIaKhaH7sAz51vBjuQhTnD3zc3A9d5_-Daa5gcQCuI/edit?usp=sharing)

Here is the list of functions which were added.

#### Puppeteer (+25 Functions)
* _locateCheckable
* _locateClickable
* _locateFields
* closeOtherTabs
* dragAndDrop
* grabBrowserLogs
* grabCssPropertyFrom
* grabHTMLFrom
* grabNumberOfVisibleElements
* grabSource
* rightClick
* scrollPageToBottom
* scrollPageToTop
* scrollTo
* seeAttributesOnElements
* seeCssPropertiesOnElements
* seeInField
* seeNumberOfElements
* seeNumberOfVisibleElements
* seeTextEquals
* seeTitleEquals
* switchTo
* waitForInvisible
* waitInUrl
* waitUrlEquals

#### Protractor (+8 Functions)
* closeCurrentTab
* grabSource
* openNewTab
* seeNumberOfElements
* seeTextEquals
* seeTitleEquals
* switchToNextTab
* switchToPreviousTab

#### Nightmare (+1 Function)
* waitForInvisible

### Improvement to WebDriverIO Helper reliability
* Refactoring WebDriverIO to make serial WebDriverIO calls instead of in parallel due to issue with chromedriver and the number of accepted parallel requests. Now no retries are required when running the WebDriverIO Helper tests!
* Fixed tests which were incorrectly passing due to above issue (for example I.seeInField)

#### Background and supporting materials

I found this issue on my build setup when trying to submit other PRs. So when running my tests I would get sporadic failures on some tests (seemingly random).

webdriverio-flakeytests

After some research I found the following issues in the Chromedriver and the WebDriverIO repos:

* [chrome driver issue](https://bugs.chromium.org/p/chromedriver/issues/detail?id=2152#c9)
* [Implement a request queue to avoid having to many parallel requests #2406](https://github.com/webdriverio/webdriverio/issues/2406)
* [I updated the selenium web driver and suddenly one test is failing. #2262](https://github.com/webdriverio/webdriverio/issues/2262)

After refactoring the code, all of the tests passed with no problems. I confirmed the results by removing the retry setting on the mocha tests and then repeating each test 10 times and checking for any failures, however it seems to have reduced the flakiness. But time will tell for sure.

### Bug Fixes

* [WebDriverIO] Various Tests which were incorrectly passing eventhough the should not have been (i.e. I.seeTextEquals and 1 or 2 more...)
* [WebDriverIO] Fixing resizeWindow when resizing to 'maximize' which was previously not working. Also added unit test
* [WebDriverIO] Fixing resizing window to full screen when taking a screenshot
* [Mochawesome] Added null checking in _before hook which caused the codeceptjs shell mode to fail. Fixes #896
* [docs] Added option about supporting the --invert option with --grep
* [assert/equal] Added .message property as it is often used in the Helper unit tests and is a standard property to access for exceptions (I think).

## Review required

@DavertMik

### WebDriverIO and Puppeteer `proceedSeeInField` Function

For some reason the functions `proceedSeeInField` in Puppeteer and `proceedSeeField` in WebDriverIO did my head in. The code is definately not my best, and probably needs to be refactored before merging the pull request.
At the moment the code passed all of the `check fields: #seeInField, #seeCheckboxIsChecked, ...` tests, but I must admit I was only basing how the function should behave by looking at the unit tests...and the unit tests are that clear.

I believe it was by accident that the previous tests were passing, as I was seeing unhandle exceptions and the tests were still passing.

### CODE REVIEW

**Comment by DavertMik** [comment]
Wow, looks awesome. Thanks! 
I will need to take a deeper look to review this!

**Comment by reubenmiller** [comment]
I just pushed another commit for `I.scrollTo`. Though I didn't manage to write a test for it because I can't decide how to it and the webdriver io helper doesn't have a test for it at the moment...so at least it is on par ;)

I'll update the PR description with the two new helper functions (I.scrollTo and I.dragAndDrop) that I did today.

**Comment by DavertMik** [comment]
Hey, @reubenmiller please join the development chat in Slack https://join.slack.com/t/codeceptjs/shared_invite/enQtMzA4ODQ1MTExMjE2LTE4ZjNkN2EwNjgxOWZjMWQ0NmQzOGVmMTg5NzEzMDdlYzAyNGZjNzlmMGExMjA3OWVlNTc2NmM3ZTY0NWJmYzA

I just started it so not to big crowd there but it's better to have it later than never

File: lib/helper/Protractor.js [hunk 1]
```diff
@@ -210,7 +210,10 @@ class Protractor extends Helper {
     if (!this.options.keepCookies) {
       await this.browser.manage().deleteAllCookies();
     }
-    await this.browser.executeScript('localStorage.clear();');
+    const url = await this.browser.getCurrentUrl();
+    if (!/data:,/i.test(url)) {
```
Comment (DavertMik) [inline lib/helper/Protractor.js]: :+1:

**Comment by DavertMik** [comment]
> For some reason the functions proceedSeeInField in Puppeteer and proceedSeeField in WebDriverIO did my head in. The code is definately not my best, and probably needs to be refactored before merging the pull request.

Maybe but I do not have any good suggestions, anyway :(

File: lib/helper/Protractor.js (hunk 1 above)
Comment (reubenmiller) [reply lib/helper/Protractor.js]: Yeah I had this code in my project because I had run into the problem before in chrome, so one less corner case :)

**Pull request merged**